### PR TITLE
[Gale Arabic] Added grapheme based AM and support nnet3 and chain modeling.

### PR DESCRIPTION
**Pull request by amali**
The main difference; s5 is using phoneme for modeling, while s5b is using grapheme, which is easier in Arabic ASR for few reasons:
1- Given that Arabic is morphologically rich language, we need large lexicon to achieve reasonable OOV. So, using grapheme helps to achieve this with small degradation in WER.
2- Grapheme system doesn't depend on any NLP which make it easy to maintain.
3- Grapheme system is easier to adapt the current (standard Arabic) models for dialectal Arabic, which is under-resourced in terms of being able to get phoneme representation.
Also, the s5b has the latest modeling in nnet3 and chain tdnn lstm (I started from tedlium recipe)
One motivation to support both phoneme and grapheme based is that for cases like dialectal identification, we found phoneme recognition is useful, so I would recommend to maintain both.
Actually, I started by making the modification in the same s5, but I thought it may be useful to keep both approaches.
What do you think?

### CODE REVIEW

**Comment by danpovey** [comment]
Thanks for the PR.
Can you summarize how the "s5b" recipe differs from the "s5" recipe?  It's
possible that we may ask you to just modify the existing "s5" recipe-- it's
under unusual circumstances that we create a newly-numbered recipe.
Of course, it would also be easier to see your changes if they were
in-place.  But first tell us how it differs from the existing recipe.

**Comment by danpovey** [comment]
OK, I guess it's reasonable to have a different experiment number.
But I think it's in need of a bit of cleanup-- can you remove the nnet and nnet2 experiment scripts, unless you think one of them is important for some reason, and also make sure the sgmm script is commented as it appears in the run.sh?  And remove any configuration files that would no longer be needed as a result, like fbank.conf?  Also go through all added files in the PR and make sure they are all called from somewhere [delete orphan files].

**Comment by amali** [comment]
@danpovey, I pushed the recommended changes.
I wonder if it makes sense to link common scripts under local across different experiments to avoid maintaining more than one version of the same script, like split_wer.sh?

File: egs/gale_arabic/s5b/conf/online_pitch.conf [hunk 1]
```diff
@@ -0,0 +1,46 @@
+## This config is given by conf/make_pitch_online.sh to the program compute-and-process-kaldi-pitch-feats,
+## and is copied by steps/online/nnet2/prepare_online_decoding.sh and similar scripts, to be given
```
Comment (danpovey) [inline egs/gale_arabic/s5b/conf/online_pitch.conf]: this shouldn't be needed if pitch features aren't being used, and I'd be surprised if they are being used for arabic (search for make_mfcc_pitch in local/)

Comment (amali) [inline egs/gale_arabic/s5b/conf/online_pitch.conf]: @danpovey, pushed.
I am not sure if other configurations like decode_dnn.config is used by nnet3/chain or no. Not called  by steps/nnet3/decode.sh or any of the local scripts.

**Comment by danpovey** [comment]
unless explicitly passed in by a script in local/, it is not used-- you can
remove it.

**Comment by danpovey** [comment]
@keli78, can you please check out this PR and try to run it on our grid?

**Comment by keli78** [comment]
Sure. Will do that soon.

**Comment by keli78** [comment]
I run the run.sh in gale_arabic/s5b and got warning messages as below:
...
Bad perplexities   . at ./../../../tools/kaldi_lm/optimize_alpha.pl line 30.
No such file data/local/lm/3gram-mincount/lm_unpruned.gz
steps/make_mfcc.sh --cmd queue.pl --mem 2G --nj 120 data/train exp/make_mfcc/train mfcc
utils/validate_data_dir.sh: empty file spk2utt
Empty list of recordings (bad file data/train/segments)?
steps/compute_cmvn_stats.sh data/train exp/make_mfcc/train mfcc
make_cmvn.sh: no such file data/train/feats.scp
steps/make_mfcc.sh --cmd queue.pl --mem 2G --nj 120 data/test exp/make_mfcc/test mfcc
utils/validate_data_dir.sh: empty file spk2utt
Empty list of recordings (bad file data/test/segments)?
steps/compute_cmvn_stats.sh data/test exp/make_mfcc/test mfcc
make_cmvn.sh: no such file data/test/feats.scp
subset_data_dir.sh: cannot subset to more utterances than you originally had.

**Comment by amali** [comment]
@keli78, I guess, this is because you didn't point to the data on your cluster:
Please modify this section in run.sh to look for the LDC data on your grid:
audio=(
  /data/sls/scratch/amali/data/GALE/LDC2013S02
  /data/sls/scratch/amali/data/GALE/LDC2013S07 
  /data/sls/scratch/amali/data/GALE/LDC2014S07 
)
text=(
  /data/sls/scratch/amali/data/GALE/LDC2013T17.tgz 
  /data/sls/scratch/amali/data/GALE/LDC2013T04.tgz 
  /data/sls/scratch/amali/data/GALE/LDC2014T17.tgz
)
I just checked it now, and it work fine .... If you still have error please run it with bash -x and send me the log file, and I will check again.

**Comment by danpovey** [comment]
The script should be a bit more careful about checking for errors and
exiting is soon as an error is found.  E.g. do the data preparation scripts
return with status 1 if there is an error?  Does the run.sh fail if the
data preparation fails?  [does it either do set -e, or do '|| exit 1' after
each command?]

**Comment by amali** [comment]
Good point. Data preparation was copied from the s5b. Now added the bash -e and exit if the data preparation fails.

**Comment by keli78** [comment]
I checked out the newest version, changed the data directories, and rerun the script in s5b. It works fine now. It's in the mono phone training stage and looks fine except for some warning messages: 
WARNING (gmm-init-model[5.0.16~1-65bea]:InitAmGmm():gmm-init-model.cc:55) Tree has pdf-id 15 with no stats; corresponding phone list: 62 63 64 65
WARNING (gmm-init-model[5.0.16~1-65bea]:InitAmGmm():gmm-init-model.cc:55) Tree has pdf-id 36 with no stats; corresponding phone list: 146 147 148 149
** The warnings above about 'no stats' generally mean you have phones **
** (or groups of phones) in your phone set that had no corresponding data. **
** You should probably figure out whether something went wrong, **
** or whether your data just doesn't happen to have examples of those **
** phones. **

Not sure whether this is a big issue.

**Comment by danpovey** [comment]
Ke, check the phones.txt and figure out what phones those correspond to.

**Comment by keli78** [comment]
62 63 64 65 are a_B, a_E, a_I, a_S respectively and 146 147 148 149 are {_B, {_E, {_I, {_S respectively

**Comment by amali** [comment]
Thanks for spotting this. "{" is very rare in the lexicon and should not be used in the grapheme mapping. I also removed "a" and make sure no vowelization in the data. The original lexicon needs a bit of tidying up.
In general, I think it wouldn't matter to the final results.

**Comment by keli78** [comment]
Got an error: steps/online/nnet2/extract_ivectors_online.sh: No such file data/dev_hires/feats.scp
I found that the dev data seems not be processed : Line 73 in run.sh: for x in train test ; do

**Comment by danpovey** [comment]
@amali, have you addressed the latest issue that Ke found, with the crash?
I don't know whether there were still other outstanding issues with this.

**Comment by amali** [comment]
@keli78, I am not using dev set in the new recipe. Also, I am using only the nnet3 scripts. Which part of the code is calling steps/online/nnet2/extract_ivectors_online.sh?

**Comment by danpovey** [comment]
Grep for that in local/nnet3.  Maybe a script there wrongly refers to dev.

File: egs/gale_arabic/s5b/local/gale_data_prep_txt.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+     tar -xvf $cdx
+  elif [  -d "$cdx" ]; then
+    ln -s $cdx `basename $cdx`
+  else
+    echo "I don't really know what I shall do with $cdx " >&2 
+  fi
+done
+
+find -L . -type f -name *.tdf | while read file; do 
+sed '1,3d' $file
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/gale_data_prep_txt.sh]: please add a comment saying what this sed command does, most people only use it for the s command.

File: egs/gale_arabic/s5b/local/gale_data_prep_txt.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  if [[ $cdx  == *.tgz ]] ; then
+     tar -xvf $cdx
+  elif [  -d "$cdx" ]; then
+    ln -s $cdx `basename $cdx`
+  else
+    echo "I don't really know what I shall do with $cdx " >&2 
+  fi
+done
+
+find -L . -type f -name *.tdf | while read file; do
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/gale_data_prep_txt.sh]: You need to quote *.tdf.

File: egs/gale_arabic/s5b/local/gale_data_prep_txt.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+perl ${top_pwd}/local/normalize_transcript_BW.pl contentall.tmp$$ contentall.buck.tmp$$
+
+paste allid.tmp$$ contentall.buck.tmp$$ | sed 's: $::' | awk '{if (NF>5) {print $0}}'  > all_1.tmp$$
+
+awk '{$1="";print $0}' all_1.tmp$$ | sed 's:^ ::' > $galeData/all
+awk '{if ($1 == "report") {$1="";print $0}}' all_1.tmp$$ | sed 's:^ ::' >  $galeData/report
+awk '{if ($1 == "conversational") {$1="";print $0}}' all_1.tmp$$ | sed 's:^ ::' > $galeData/conversational
+
+cd ..;
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/gale_data_prep_txt.sh]: I don't see where you're removing the temporary files.

File: egs/gale_arabic/s5b/local/gale_format_data.sh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+echo  "Checking how stochastic G is (the first of these numbers should be small):"
+fstisstochastic data/lang_test/G.fst
+
+## Check lexicon.
+## just have a look and make sure it seems sane.
+echo "First few lines of lexicon FST:"
+fstprint   --isymbols=data/lang/phones.txt --osymbols=data/lang/words.txt data/lang/L.fst  | head
+
+echo Performing further checks
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/gale_format_data.sh]: Please make sure that all echo statements of this nature start with the name of the script, e.g.:
`echo "$0: Performing further checks"`

File: egs/gale_arabic/s5b/local/gale_prep_grapheme_dict.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+# silence phones, one per line.
+echo SIL > $dir/silence_phones.txt
+echo SIL > $dir/optional_silence.txt
+
+# nonsilence phones; on each line is a list of phones that correspond
+# really to the same base phone.
+cat tmp2.$$ | tr -s ' ' '\n' | grep -v ^$  | sort -u >  $dir/nonsilence_phones.txt || exit 1;
+
+sed -i '1i<UNK> SIL' $dir/lexicon.txt
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/gale_prep_grapheme_dict.sh]: again, please explain what this command is doing.

File: egs/gale_arabic/s5b/local/gale_train_lms.sh [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+
+lexicon=data/local/dict/lexicon.txt 
+[ ! -f $lexicon ] && echo "$0: No such file $lexicon" && exit 1;
+
+# check if sri is installed or no
+sri_installed=false
+which ngram-count  &>/dev/null
+if [[ $? == 0 ]]; then 
+sri_installed=true
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/gale_train_lms.sh]: indent this line please.

File: egs/gale_arabic/s5b/local/gale_train_lms.sh [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+ ngram-count -text $sdir/train -order 3 -limit-vocab -vocab $sdir/wordlist -unk \
+   -map-unk "<UNK>" -kndiscount -interpolate -lm $sdir/srilm.o3g.kn.gz
+ ngram -lm $sdir/srilm.o3g.kn.gz -ppl $sdir/heldout 
+# 0 zeroprobs, logprob= -250954 ppl= 90.5091 ppl1= 132.482
+
+# Note: perplexity SRILM gives to Kaldi-LM model is same as kaldi-lm reports above.
+# Difference in WSJ must have been due to different treatment of <UNK>.
+ ngram -lm $dir/3gram-mincount/lm_unpruned.gz  -ppl $sdir/heldout 
+# 0 zeroprobs, logprob= -250913 ppl= 90.4439 ppl1= 132.379
+fi
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/gale_train_lms.sh]: is there a reason why you need to train the LMs with both kaldi_lm or SRILM, or is one of them unnecessary?

File: egs/gale_arabic/s5b/local/nnet3/run_lstm.sh [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+train_set=train
+gmm=tri2b  # this is the source gmm-dir for the data-type of interest; it
+                  # should have alignments for the specified training data.
+num_threads_ubm=32
+nnet3_affix=_cleaned  # cleanup affix for exp dirs, e.g. _cleaned
+
+# Options which are not passed through to run_ivector_common.sh
+affix=
+common_egs_dir=
+reporting_email="amali@qf.org.qa"
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/nnet3/run_lstm.sh]: you'll want to remove this before you check things in, or you'll be getting some spam.

File: egs/gale_arabic/s5b/local/nnet3/run_lstm.sh [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+for f in $train_data_dir/feats.scp $train_ivector_dir/ivector_online.scp \
+     $graph_dir/HCLG.fst $ali_dir/ali.1.gz $gmm_dir/final.mdl; do
+  [ ! -f $f ] && echo "$0: expected file $f to exist" && exit 1
+done
+
+if [ $stage -le 12 ]; then
+  echo "$0: creating neural net configs"
+  config_extra_opts=()
+  [ ! -z "$lstm_delay" ] && config_extra_opts+=(--lstm-delay "$lstm_delay")
+  steps/nnet3/lstm/make_configs.py  "${config_extra_opts[@]}" \
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/nnet3/run_lstm.sh]: This script (make_configs.py) is deprecated.  We should aim to replace this with a newer xconfig-based recipe soon.  But we can do that after this version of the recipe is checked in.  Same applies to the TDNN script.

File: egs/gale_arabic/s5b/local/score_combine.sh [hunk 11]
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
+
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/score_combine.sh]: if this is not needed it can be removed.

File: egs/gale_arabic/s5b/local/score_mbr.sh [hunk 12]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+
+# Script for minimum bayes risk decoding.
+
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/score_mbr.sh]: again, if this is not needed it can be removed.. .not sure if it's ever called?

File: egs/gale_arabic/s5b/local/split_wer_per_corpus.sh [hunk 13]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
+# Report WER for reports and conversational
+# Copyright 2014 QCRI (author: Ahmed Ali)
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/split_wer_per_corpus.sh]: This seems to be an unused copy of split_wer.sh.

File: egs/gale_arabic/s5b/run.sh [hunk 14]
```diff
@@ -0,0 +1,167 @@
+#!/bin/bash -e
+
+# Copyright 2014 QCRI (author: Ahmed Ali)
+# Apache 2.0
+
+. path.sh
+. cmd.sh   ## You'll want to change cmd.sh to something that will work on your system.
+           ## This relates to the queue.
+nJobs=120
```
Comment (danpovey) [inline egs/gale_arabic/s5b/run.sh]: better to use underscores in variable names, e.g. num_jobs, num_decode_jobs.

File: egs/gale_arabic/s5b/run.sh [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+steps/decode_fmllr.sh --nj $nDecodeJobs --cmd \
+  "$decode_cmd" exp/tri3b/graph data/test exp/tri3b/decode
+
+# From 3b system, align all data.
+steps/align_fmllr.sh --nj $nJobs --cmd "$train_cmd" \
+  data/train data/lang exp/tri3b exp/tri3b_ali || exit 1;
+  
+
+# nnet3 cross-entropy 
+local/nnet3/run_tdnn.sh #tdnn recipe:
```
Comment (danpovey) [inline egs/gale_arabic/s5b/run.sh]: I'd recommend to have most of these commented out in the run.sh by default, and just run the best of these recipes.

File: egs/gale_arabic/s5b/local/gale_prep_grapheme_dict.sh [hunk 16]
```diff
@@ -16,7 +16,7 @@ bzcat ar-ar_grapheme_lexicon_2016-02-09.bz2  | sed '1,3d' | awk '{print $1}'  >
 bzcat ar-ar_lexicon_2014-03-17.txt.bz2 | sed '1,3d' | awk '{print $1}' >>  tmp$$
 # (2) Now we add all the words appeared in the training data
 cat data/local/train/text | cut -d ' ' -f 2- | tr -s " " "\n" | sort -u >> tmp$$
-grep -v [0-9] tmp$$ |  sort -u > tmp1.$$
+grep -v [0-9] tmp$$ |  sed -e 's:[FNKaui\~o\`]::g' -e 's:{:}:g' | sort -u > tmp1.$$ # remove vowels and  rare alef wasla
```
Comment (keli78) [inline egs/gale_arabic/s5b/local/gale_prep_grapheme_dict.sh]: Got some errors as below: 
328 Checking data/local/dict/lexiconp.txt                                           
329 --> reading data/local/dict/lexiconp.txt                                        
330 --> ERROR: phone "a" is not in {, non}silence.txt (line 21515)                  
331 --> ERROR: phone "a" is not in {, non}silence.txt (line 33745)                  
332 --> ERROR: phone "a" is not in {, non}silence.txt (line 455372)                 
333 --> ERROR: phone "a" is not in {, non}silence.txt (line 455373)                 
334 --> ERROR: phone "a" is not in {, non}silence.txt (line 455374)                 
335 --> ERROR: phone "a" is not in {, non}silence.txt (line 455375)                 
336 --> ERROR: phone "a" is not in {, non}silence.txt (line 455376)                 
337 --> ERROR: phone "a" is not in {, non}silence.txt (line 630905)                 
338 --> ERROR: phone "{" is not in {, non}silence.txt (line 955030)                 
339 --> ERROR: phone "{" is not in {, non}silence.txt (line 955031)                 
340 --> ERROR: phone "{" is not in {, non}silence.txt (line 955032)                 
341                                                                                 
342 Checking lexicon pair data/local/dict/lexicon.txt and data/local/dict/lexiconp.txt
343 --> ERROR: data/local/dict/lexicon.txt and data/local/dict/lexiconp.txt mismatch at line 21515. sorting?
344 Checking data/local/dict/extra_questions.txt ...                                
345 --> data/local/dict/extra_questions.txt is empty (this is OK)                   
346 --> ERROR validating dictionary directory data/local/dict (see detailed error messages above)
347                                                                                 
348 *Error validating directory data/local/dict*

File: egs/gale_arabic/s5b/local/nnet3/run_tdnn.sh [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+done
+
+
+if [ $stage -le 12 ]; then
+  if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d $dir/egs/storage ]; then
+    utils/create_split_dir.pl \
+     /export/b0{3,4,5,6}/$USER/kaldi-data/egs/tedlium-$(date +'%m_%d_%H_%M')/s5_r2/$dir/egs/storage $dir/egs/storage
+  fi
+   
+  bash -x steps/nnet3/tdnn/train.sh --stage $train_stage \
```
Comment (danpovey) [inline egs/gale_arabic/s5b/local/nnet3/run_tdnn.sh]: i think this bash -x should be removed.

File: egs/gale_arabic/s5b/local/gale_train_lms.sh (hunk 8 above)
Comment (amali) [inline egs/gale_arabic/s5b/local/gale_train_lms.sh]: I assume, it was always using both sri and kali lm and compare the perplexity to make sure performance is about the same.  I didn't check it since committing first version .

**Comment by danpovey** [comment]
I'd suggest to remove whichever of the two is not ultimately being used by
the rest of the script.  I don't like unnecessary dependencies.

**Comment by amali** [comment]
@keli78, I think the problem you got was because prepare_lang.sh will not re-generate lexiconp.txt if it exists and check it, and since I remove two phones "{" and "a", it is failing because of that. If you remove ldata/local/dict/ and restart it is ok. Ideally, I will not make changes often and hopefully it is one-off.

File: egs/gale_arabic/s5/run.sh [hunk 18]
```diff
@@ -42,25 +44,25 @@ galeData=GALE
 # By copying and pasting into your shell.
 
 #copy the audio files to local folder wav and convet flac files to wav
-local/gale_data_prep_audio.sh  "${audio[@]}" $galeData 
+local/gale_data_prep_audio.sh  "${audio[@]}" $galeData || exit 1;
 
 #get the transcription and remove empty prompts and all noise markers  
-local/gale_data_prep_txt.sh  "${text[@]}" $galeData
+local/gale_data_prep_txt.sh  "${text[@]}" $galeData || exit 1;
```
Comment (amali) [inline egs/gale_arabic/s5/run.sh]: fixed

File: egs/gale_arabic/s5b/local/gale_data_prep_txt.sh (hunk 4 above)
Comment (amali) [inline egs/gale_arabic/s5b/local/gale_data_prep_txt.sh]: rm -fr $txtdir and we keep GALE/{report, conversational}

File: egs/gale_arabic/s5b/local/gale_format_data.sh (hunk 5 above)
Comment (amali) [inline egs/gale_arabic/s5b/local/gale_format_data.sh]: done

**Comment by danpovey** [comment]
Thanks... I'll wait for @keli78 to check it again before committing (and
also for your fix to the LM-preparation script).

File: egs/gale_arabic/s5b/local/split_wer_per_corpus.sh (hunk 13 above)
Comment (amali) [inline egs/gale_arabic/s5b/local/split_wer_per_corpus.sh]: I actually have the same question, no idea where it comes from. However, calculating wer per program or per corpus may be interesting to extend it further to be per genre in cases like MGB

**Comment by danpovey** [comment]
Delete it if it's unused anywhere.
The same in the baseline script.  Probably just a mistake, someone forgot
git rm.

**Comment by amali** [comment]
All done. Only remaining one issue is local/gale_train_lms.sh builds both sri and kaldi LM. I think it is OK to keep it this way.

**Comment by danpovey** [comment]
Can't you figure  out which of the two LMs it's using later on, and then
just delete the other commands?  It doesn't seem to me very hard.  It would
have always been a mistake, to run both sets of commands.

**Comment by keli78** [comment]
Sure, I'll check it as soon as it's ready.

**Comment by keli78** [comment]
It works fine although is still running (train nnet3 tdnn stage). I only got several warnings as below: 
analyze_phone_length_stats.py: WARNING: optional-silence SIL is seen only 76.7044085762% of the time at utterance begin.  This may not be optimal.
and one warning like below: 
WARNING (feat-to-dim[5.0.20~1-7e149]:Close():kaldi-io.cc:501) Pipe utils/filter_scp.pl --exclude exp/nnet3_cleaned/tdnn_sp/egs/valid_uttlist data/train_sp_hires_comb/split6/1/feats.scp | apply-cmvn --norm-means=false --norm-vars=false --utt2spk=ark:data/train_sp_hires_comb/split6/1/utt2spk scp:data/train_sp_hires_comb/split6/1/cmvn.scp scp:- ark:     - | had nonzero return status 36096

**Comment by danpovey** [comment]
Those warnings are probably OK.

**Comment by keli78** [comment]
I think it mainly works fine except for the issue below: 
4022 local/nnet3/run_ivector_common.sh: data/train_sp_hires/feats.scp already exists.
4023  ... Please either remove it, or rerun this script with stage > 2.
Do I need to further test it?

**Comment by danpovey** [comment]
That's OK, it's supposed to fail like that if you are rerunning.

**Comment by danpovey** [comment]
Merging-- thanks!

**Comment by amali** [comment]
Thanks for the review and the help to push the new recipe!
